js/coindcx.js

```diff
@@ -0,0 +1,596 @@
+            'price': price,
+            'amount': amount,
+            'cost': price * amount,
+            'fee': this.safeFloat (trade, 'fee_amount'),
+        };
+    }
+
+    async fetchOrderBook (symbol, limit = undefined, params = {}) {
+        // https://coindcx-official.github.io/rest-api/?shell#order-book
+        if (symbol === undefined) {
+            throw new ArgumentsRequired (this.id + ' fetchOrderBook requires a `symbol` argument');
+        }
+        await this.loadMarkets ();
+        if (!(symbol in this.markets)) {
+            throw new ExchangeError (this.id + ' has no symbol ' + symbol);
+        }
+        const market = this.market (symbol);
+        const marketInfo = this.safeValue (market, 'info');
+        const coindcxPair = this.safeString (marketInfo, 'pair');
+        if (coindcxPair === undefined) {
+            throw new ExchangeError (this.id + ' has no pair (look at market\'s info) value for ' + symbol);
+        }
+        const request = {
+            'pair': coindcxPair,
+        };
+        const response = await this.publicGetMarketDataOrderbook (this.extend (request, params));
+        return this.parseOrderBookData (response);
+    }
+
+    parseOrderBookData (orderBook) {
+        const bids = this.safeValue (orderBook, 'bids', {});
+        const asks = this.safeValue (orderBook, 'asks', {});
+        return {
+            'bids': this.sortBy (this.parseBidAskData (bids), 0, true),
+            'asks': this.sortBy (this.parseBidAskData (asks), 0),
+            'timestamp': undefined,
+            'datetime': undefined,
+            'nonce': undefined,
+        };
+    }
+
+    parseBidAskData (bidsOrAsks) {
+        const priceKeys = Object.keys (bidsOrAsks);
+        const parsedData = [];
+        for (let i = 0; i < priceKeys.length; i++) {
+            const key = priceKeys[i];
+            const price = parseFloat (key);
+            const amount = parseFloat (bidsOrAsks[key]);
+            parsedData.push ([price, amount]);
+        }
+        return parsedData;
+    }
+
+    async fetchBalance (params = {}) {
+        // https://coindcx-official.github.io/rest-api/?javascript#get-balances
+        await this.loadMarkets ();
+        const request = {
+            'timestamp': this.milliseconds (),
+        };
+        const response = await this.privatePostExchangeV1UsersBalances (this.extend (request, params));
+        const result = { 'info': response };
+        for (let i = 0; i < response.length; i++) {
+            const balance = response[i];
+            const currencyId = this.safeString (balance, 'currency');
+            const code = this.safeCurrencyCode (currencyId);
+            if (!(code in result)) {
+                const account = this.account ();
+                const free = this.safeFloat (balance, 'balance');
+                const used = this.safeFloat (balance, 'locked_balance');
+                const total = parseFloat (free + used);
```

Comment:
This line is redundant, we don't need it here, because total is calculated in `parseBalance`, if missing in the exchange response. See how this is done in the other exchanges. Also, it violates the CONTRIBUTING guidelines here:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#adding-numbers-and-concatenating-strings